weather.py

```diff
@@ -0,0 +1,198 @@
+    murree_weather_month = []
+    if not os.path.isfile(weather_file_path):
+        return murree_weather_month
+
+    with open(weather_file_path, mode="r") as file:
+        csv_reader = csv.DictReader(file)
+        headers = [header.strip() for header in csv_reader.fieldnames]
+        for row in csv_reader:
+            row = {key.strip(): value for key, value in row.items()}
+            filtered_row = {key: row[key] for key in required_labels if key in headers}
+
+            murree_weather_month.append(filtered_row)
+
+    return murree_weather_month
+
+
+def load_yearly_weather_data(directory_path, year):
+    year_weather_data = []
+
+    for month in month_abbr_map.keys():
+        if not month:
+            continue
+
+        monthly_data = load_monthly_weather_data(directory_path, year, month)
+        year_weather_data.extend(monthly_data)
+
+    return year_weather_data
+
+
+def load_monthly_weather_data(directory_path, year, month):
+    murree_weather_file = f"Murree_weather_{year}_{month_abbr_map[month]}.txt"
+    full_file_path = os.path.join(directory_path, murree_weather_file)
+    return read_weather_file(full_file_path)
+
+
+def calculate_yearly_stats(weather_data):
+    highest_temperature = float("-inf")
+    lowest_temperature = float("inf")
+    highest_humidity = 0
+
+    highest_temprature_day = None
+    lowest_temprature_day = None
+    highest_humidity_day = None
+
+    for daily_data in weather_data:
+        date_key = "PKT" if "PKT" in daily_data else "PKST"
+
+        if daily_data["Max TemperatureC"]:
+            max_temp = int(daily_data["Max TemperatureC"])
+            if max_temp > highest_temperature:
+                highest_temperature = max_temp
+                highest_temprature_day = daily_data[date_key]
+
+            if max_temp < lowest_temperature:
+                lowest_temperature = max_temp
+                lowest_temprature_day = daily_data[date_key]
+
+            if daily_data["Mean Humidity"]:
+                max_humidity = int(daily_data["Mean Humidity"])
+                if max_humidity > highest_humidity:
+                    highest_humidity = max_humidity
+                    highest_humidity_day = daily_data[date_key]
+
+    return {
+        "Highest": (highest_temperature, highest_temprature_day),
+        "Lowest": (lowest_temperature, lowest_temprature_day),
+        "Humidity": (highest_humidity, highest_humidity_day)
+    }
+
+
+def calculate_monthly_stats(weather_data):
+    high_temperatures = []
+    low_temperatures = []
+    mean_humidities = []
+
+    for daily_data in weather_data:
+        if daily_data["Max TemperatureC"]:
+            high_temperatures.append(int(daily_data["Max TemperatureC"]))
+
+        if daily_data["Min TemperatureC"]:
+            low_temperatures.append(int(daily_data["Min TemperatureC"]))
+
+        if daily_data["Mean Humidity"]:
+            mean_humidities.append(int(daily_data["Mean Humidity"]))

```

Comment:
Average calculation should be on generic function which is receiving list of inputs and calculate averages on it with respect to key.
Currently this is violation of DRY Principle.